File: scripts/patchers/kernel_jb_patch_hook_cred_label.py
"""Mixin: KernelJBPatchHookCredLabelMixin."""

from .kernel_jb_base import asm, _rd32, _rd64, RET, NOP, struct

PACIBSP = bytes([0x7F, 0x23, 0x03, 0xD5])  # 0xD503237F


class KernelJBPatchHookCredLabelMixin:
    def _find_vnode_getattr_via_string(self):
        """Find vnode_getattr by locating a caller function via string ref.

        The string "vnode_getattr" appears in format strings like
        "%s: vnode_getattr: %d" inside functions that CALL vnode_getattr.
        We find such a caller, then extract the BL target near the string
        reference to get the real vnode_getattr address.

        Previous approach: find_string → find_string_refs → find_function_start
        was wrong because it returned the CALLER (e.g. an AppleImage4 function)
        instead of vnode_getattr itself.
        """
        str_off = self.find_string(b"vnode_getattr")
        if str_off < 0:
            return -1

        refs = self.find_string_refs(str_off)
        if not refs:
            return -1

        # The string ref is inside a function that calls vnode_getattr.
        # Scan backward from the string ref for a BL instruction — the
        # nearest preceding BL is very likely the BL vnode_getattr call
        # (the error message prints right after the call fails).
        ref_off = refs[0][0]  # ADRP offset
        for scan_off in range(ref_off - 4, ref_off - 64, -4):
            if scan_off < 0:
                break
            insn = _rd32(self.raw, scan_off)
            if (insn >> 26) == 0x25:  # BL opcode
                imm26 = insn & 0x3FFFFFF
                if imm26 & (1 << 25):
                    imm26 -= 1 << 26  # sign extend
                target = scan_off + imm26 * 4
                if any(s <= target < e for s, e in self.code_ranges):
                    self._log(
                        f"  [+] vnode_getattr at 0x{target:X} "
                        f"(via BL at 0x{scan_off:X}, "
                        f"near string ref at 0x{ref_off:X})"
                    )
                    return target

        # Fallback: try additional string hits
        start = str_off + 1
        for _ in range(5):
            str_off2 = self.find_string(b"vnode_getattr", start)
            if str_off2 < 0:
                break
            refs2 = self.find_string_refs(str_off2)
            if refs2:
                ref_off2 = refs2[0][0]
                for scan_off in range(ref_off2 - 4, ref_off2 - 64, -4):
                    if scan_off < 0:
                        break
                    insn = _rd32(self.raw, scan_off)
                    if (insn >> 26) == 0x25:  # BL
                        imm26 = insn & 0x3FFFFFF
                        if imm26 & (1 << 25):
                            imm26 -= 1 << 26
                        target = scan_off + imm26 * 4
                        if any(s <= target < e for s, e in self.code_ranges):
                            self._log(
                                f"  [+] vnode_getattr at 0x{target:X} "
                                f"(via BL at 0x{scan_off:X})"
                            )
                            return target
            start = str_off2 + 1

        return -1

    def patch_hook_cred_label_update_execve(self):
        """Inline-trampoline the sandbox cred_label_update_execve hook.

        Injects ownership-propagation shellcode by replacing the first
        instruction (PACIBSP) of the original hook with ``B cave``.
        The cave runs PACIBSP, performs vnode_getattr ownership propagation,
        then ``B hook+4`` to resume the original function.

        Previous approach (ops table pointer rewrite) broke the chained
        fixup integrity check, causing PAC failures in unrelated kexts.
        Inline trampoline avoids PAC entirely — B is PC-relative.
        """
        self._log(
            "\n[JB] _hook_cred_label_update_execve: "
            "inline trampoline + shellcode"
        )

        # ── 1. Find vnode_getattr via string anchor ──────────────
        vnode_getattr_off = self._resolve_symbol("_vnode_getattr")
        if vnode_getattr_off < 0:
            vnode_getattr_off = self._find_vnode_getattr_via_string()

        if vnode_getattr_off < 0:
            self._log("  [-] vnode_getattr not found")
            return False

        # ── 2. Find sandbox ops table ────────────────────────────
        ops_table = self._find_sandbox_ops_table_via_conf()
        if ops_table is None:
            self._log("  [-] sandbox ops table not found")
            return False

        # ── 3. Find hook index dynamically ───────────────────────
        # mpo_cred_label_update_execve is one of the largest sandbox
        # hooks at an early index (< 30).  Scan for it.
        hook_index = -1
        orig_hook = -1
        best_size = 0
        for idx in range(0, 30):
            entry = self._read_ops_entry(ops_table, idx)
            if entry is None or entry <= 0:
                continue
            if not any(s <= entry < e for s, e in self.code_ranges):
                continue
            fend = self._find_func_end(entry, 0x2000)
            fsize = fend - entry
            if fsize > best_size:
                best_size = fsize
                hook_index = idx
                orig_hook = entry

        if hook_index < 0 or best_size < 1000:
            self._log(
                "  [-] hook entry not found in ops table "
                f"(best: idx={hook_index}, size={best_size})"
            )
            return False

        self._log(
            f"  [+] hook at ops[{hook_index}] = 0x{orig_hook:X} "
            f"({best_size} bytes)"
        )

        # Verify first instruction is PACIBSP
        first_insn = self.raw[orig_hook : orig_hook + 4]
        if first_insn != PACIBSP:
            self._log(
                f"  [-] first insn not PACIBSP "
                f"(got 0x{_rd32(self.raw, orig_hook):08X})"
            )
            return False

        # ── 4. Find code cave ────────────────────────────────────
        cave = self._find_code_cave(200)
        if cave < 0:
            self._log("  [-] no code cave found")
            return False
        self._log(f"  [+] code cave at 0x{cave:X}")

        # ── 5. Encode branches ─────────────────────────────────
        # BL cave→vnode_getattr  (slot 18)
        vnode_bl_off = cave + 18 * 4
        vnode_bl = self._encode_bl(vnode_bl_off, vnode_getattr_off)
        if not vnode_bl:
            self._log("  [-] BL to vnode_getattr out of range")
            return False

        # B cave→hook+4  (back to STP after PACIBSP, last slot)
        b_resume_slot = 45
        b_resume_off = cave + b_resume_slot * 4
        b_resume = self._encode_b(b_resume_off, orig_hook + 4)
        if not b_resume:
            self._log("  [-] B to hook+4 out of range")
            return False

        # B hook→cave  (replaces PACIBSP at function entry)
        b_to_cave = self._encode_b(orig_hook, cave)
        if not b_to_cave:
            self._log("  [-] B to cave out of range")
            return False

        # ── 6. Build shellcode ───────────────────────────────────
        # MAC hook args: x0=old_cred, x1=new_cred, x2=proc, x3=vp
        #
        # The cave starts with PACIBSP (relocated from hook entry),
        # then performs ownership propagation, then resumes the
        # original function at hook+4 (the STP instruction).
        #
        # struct vfs_context { thread_t vc_thread; kauth_cred_t vc_ucred; }
        # Built on the stack at [sp, #0x70].
        parts = []
        parts.append(PACIBSP)                        # 0: relocated from hook
        parts.append(asm("cbz x3, #0xb0"))           # 1: if vp==NULL → slot 45
        parts.append(asm("sub sp, sp, #0x400"))      # 2
        parts.append(asm("stp x29, x30, [sp]"))      # 3
        parts.append(asm("stp x0, x1, [sp, #16]"))   # 4
        parts.append(asm("stp x2, x3, [sp, #32]"))   # 5
        parts.append(asm("stp x4, x5, [sp, #48]"))   # 6
        parts.append(asm("stp x6, x7, [sp, #64]"))   # 7
        # Construct vfs_context inline
        parts.append(asm("mrs x8, tpidr_el1"))       # 8: current_thread
        parts.append(asm("stp x8, x0, [sp, #0x70]")) # 9: {thread, cred}
        parts.append(asm("add x2, sp, #0x70"))       # 10: ctx = &vfs_ctx
        # Setup vnode_getattr(vp, &vattr, ctx)
        parts.append(asm("ldr x0, [sp, #0x28]"))     # 11: x0 = vp (saved x3)
        parts.append(asm("add x1, sp, #0x80"))       # 12: x1 = &vattr
        parts.append(asm("mov w8, #0x380"))           # 13: vattr size
        parts.append(asm("stp xzr, x8, [x1]"))       # 14: init vattr
        parts.append(asm("stp xzr, xzr, [x1, #0x10]"))  # 15: init vattr+16
        parts.append(NOP)                             # 16
        parts.append(NOP)                             # 17
        parts.append(vnode_bl)                        # 18: BL vnode_getattr
        # Check result + propagate ownership
        parts.append(asm("cbnz x0, #0x4c"))          # 19: error → slot 38
        parts.append(asm("mov w2, #0"))              # 20: changed = 0
        parts.append(asm("ldr w8, [sp, #0xCC]"))     # 21: va_mode
        parts.append(bytes([0xA8, 0x00, 0x58, 0x36]))  # 22: tbz w8,#11
        parts.append(asm("ldr w8, [sp, #0xC4]"))     # 23: va_uid
        parts.append(asm("ldr x0, [sp, #0x18]"))     # 24: new_cred
        parts.append(asm("str w8, [x0, #0x18]"))     # 25: cred->uid
        parts.append(asm("mov w2, #1"))              # 26: changed = 1
        parts.append(asm("ldr w8, [sp, #0xCC]"))     # 27: va_mode
        parts.append(bytes([0xA8, 0x00, 0x50, 0x36]))  # 28: tbz w8,#10
        parts.append(asm("mov w2, #1"))              # 29: changed = 1
        parts.append(asm("ldr w8, [sp, #0xC8]"))     # 30: va_gid
        parts.append(asm("ldr x0, [sp, #0x18]"))     # 31: new_cred
        parts.append(asm("str w8, [x0, #0x28]"))     # 32: cred->gid
        parts.append(asm("cbz w2, #0x14"))           # 33: if !changed → slot 38
        parts.append(asm("ldr x0, [sp, #0x20]"))     # 34: proc
        parts.append(asm("ldr w8, [x0, #0x454]"))    # 35: p_csflags
        parts.append(asm("orr w8, w8, #0x100"))      # 36: CS_VALID
        parts.append(asm("str w8, [x0, #0x454]"))    # 37: store
        # Restore and resume
        parts.append(asm("ldp x0, x1, [sp, #16]"))   # 38
        parts.append(asm("ldp x2, x3, [sp, #32]"))   # 39
        parts.append(asm("ldp x4, x5, [sp, #48]"))   # 40
        parts.append(asm("ldp x6, x7, [sp, #64]"))   # 41
        parts.append(asm("ldp x29, x30, [sp]"))      # 42
        parts.append(asm("add sp, sp, #0x400"))       # 43
        parts.append(NOP)                             # 44
        parts.append(b_resume)                        # 45: B hook+4

        for i, part in enumerate(parts):
            self.emit(
                cave + i * 4,
                part,
                f"shellcode+{i * 4} [_hook_cred_label_update_execve]",
            )

        # ── 7. Patch function entry ─────────────────────────────
        # Replace PACIBSP with B cave (inline trampoline).
        # No ops table modification — avoids chained fixup integrity issues.
        self.emit(
            orig_hook,
            b_to_cave,
            "B cave [_hook_cred_label_update_execve trampoline]",
        )

        return True

File: sources/FirmwarePatcher/Kernel/JBPatches/KernelJBPatchHookCredLabel.swift
// KernelJBPatchHookCredLabel.swift — JB kernel patch: Faithful upstream C23 hook
//
// Historical note: derived from the legacy Python firmware patcher during the Swift migration.
//
// Strategy (faithful upstream C23): Redirect mac_policy_ops[18]
// (_hook_cred_label_update_execve sandbox wrapper) to a code cave that:
//   1. Saves all argument registers + frame.
//   2. Calls vfs_context_current() to get the vfs context.
//   3. Calls vnode_getattr(vp, vap, ctx) to get owner/mode attributes.
//   4. If VSUID or VSGID bits set: copies owner uid/gid into the pending
//      credential and sets P_SUGID.
//   5. Restores all registers and branches back to the original sandbox wrapper.
//
// The ops[18] entry is an auth-rebase chained pointer. We re-encode it
// preserving PAC metadata but changing the target to our cave address.

import Capstone
import Foundation

extension KernelJBPatcher {
    // MARK: - Constants

    private static let hookCredLabelIndex = 18
    private static let c23CaveWords = 46 // Must match Python _C23_CAVE_WORDS

    // Expected shape of vfs_context_current prologue (5 words).
    // Python: _VFS_CONTEXT_CURRENT_SHAPE
    private static let vfsContextCurrentShape: [UInt32] = [
        ARM64.pacibspU32, // pacibsp
        ARM64.stpFP_LR_pre, // stp x29, x30, [sp, #-0x10]!
        ARM64.movFP_SP, // mov x29, sp
        ARM64.mrs_x0_tpidr_el1, // mrs x0, tpidr_el1
        ARM64.ldr_x1_x0_0x3e0, // ldr x1, [x0, #0x3e0]
    ]

    // MARK: - Entry Point

    /// Faithful upstream C23: redirect ops[18] to a vnode-getattr trampoline.
    @discardableResult
    func patchHookCredLabelUpdateExecve() -> Bool {
        log("\n[JB] _hook_cred_label_update_execve: faithful upstream C23")

        // 1. Find sandbox ops[18] entry and current wrapper target.
        guard let (opsTable, entryOff, entryRaw, wrapperOff) = findHookCredLabelWrapper() else {
            return false
        }

        // 2. Find vfs_context_current by prologue shape scan.
        let vfsCtxOff = findVfsContextCurrentByShape()
        guard vfsCtxOff >= 0 else {
            log("  [-] vfs_context_current not resolved")
            return false
        }

        // 3. Find vnode_getattr by BL scan near its log string.
        let vnodeGetattrOff = findVnodeGetattrViaString()
        guard vnodeGetattrOff >= 0 else {
            log("  [-] vnode_getattr not resolved")
            return false
        }

        // 4. Allocate code cave for 46 instructions (184 bytes).
        let caveSize = Self.c23CaveWords * 4
        guard let caveOff = findCodeCave(size: caveSize) else {
            log("  [-] no executable code cave found for faithful C23 (\(caveSize) bytes)")
            return false
        }

        // 5. Build the C23 shellcode.
        guard let caveBytes = buildC23Cave(
            caveOff: caveOff,
            vfsContextCurrentOff: vfsCtxOff,
            vnodeGetattrOff: vnodeGetattrOff,
            wrapperOff: wrapperOff
        ) else {
            log("  [-] failed to encode faithful C23 branch/call relocations")
            return false
        }

        // 6. Retarget ops[18] to cave.
        guard let newEntry = encodeAuthRebaseLike(origVal: entryRaw, targetFoff: caveOff) else {
            log("  [-] failed to encode hook ops entry retarget")
            return false
        }

        emit(entryOff, newEntry,
             patchID: "jb.hook_cred_label.ops_retarget",
             description: "retarget ops[\(Self.hookCredLabelIndex)] to faithful C23 cave [_hook_cred_label_update_execve]")

        emit(caveOff, caveBytes,
             patchID: "jb.hook_cred_label.c23_cave",
             description: "faithful upstream C23 cave (vnode getattr -> uid/gid/P_SUGID fixup -> wrapper)")

        _ = opsTable
        return true
    }

    // MARK: - Sandbox Ops Table Finder

    /// Find the sandbox mac_policy_ops table via the mac_policy_conf struct.
    /// Mirrors Python `_find_sandbox_ops_table_via_conf()`: locates the conf
    /// struct by its "Sandbox" + "Seatbelt sandbox policy" string pair, then
    /// reads the mpc_ops pointer at conf+32.
    private func findSandboxOpsTable() -> Int? {
        guard let seatbeltOff = buffer.findString("Seatbelt sandbox policy") else {
            log("  [-] Sandbox/Seatbelt strings not found")
            return nil
        }
        // Find "\0Sandbox\0" and return offset of 'S'
        guard let sandboxPattern = "\u{0}Sandbox\u{0}".data(using: .utf8),
              let sandboxRange = buffer.data.range(of: sandboxPattern)
        else {
            log("  [-] Sandbox string not found")
            return nil
        }
        let sandboxOff = sandboxRange.lowerBound + 1 // skip leading NUL

        // Collect __DATA_CONST and __DATA segment ranges.
        var dataRanges: [(Int, Int)] = []
        for seg in segments {
            if seg.name == "__DATA_CONST" || seg.name == "__DATA", seg.fileSize > 0 {
                let s = Int(seg.fileOffset)
                dataRanges.append((s, s + Int(seg.fileSize)))
            }
        }

        for (dStart, dEnd) in dataRanges {
            var i = dStart
            while i <= dEnd - 40 {
                defer { i += 8 }
                let val = buffer.readU64(at: i)
                if val == 0 || (val & (1 << 63)) != 0 { continue }
                guard (val & 0x7FF_FFFF_FFFF) == UInt64(sandboxOff) else { continue }

                let val2 = buffer.readU64(at: i + 8)
                if (val2 & (1 << 63)) != 0 { continue }
                guard (val2 & 0x7FF_FFFF_FFFF) == UInt64(seatbeltOff) else { continue }

                let valOps = buffer.readU64(at: i + 32)
                if (valOps & (1 << 63)) == 0 {
                    let opsOff = Int(valOps & 0x7FF_FFFF_FFFF)
                    log("  [+] mac_policy_conf at foff 0x\(String(format: "%X", i)), mpc_ops -> 0x\(String(format: "%X", opsOff))")
                    return opsOff
                }
            }
        }

        log("  [-] mac_policy_conf not found")
        return nil
    }

    /// Find the sandbox ops[18] wrapper, returning (opsTable, entryOff, entryRaw, wrapperOff).
    private func findHookCredLabelWrapper() -> (Int, Int, UInt64, Int)? {
        guard let opsTable = findSandboxOpsTable() else {
            log("  [-] sandbox ops table not found")
            return nil
        }

        let entryOff = opsTable + Self.hookCredLabelIndex * 8
        guard entryOff + 8 <= buffer.count else {
            log("  [-] hook ops entry outside file")
            return nil
        }

        let entryRaw = buffer.readU64(at: entryOff)
        guard entryRaw != 0 else {
            log("  [-] hook ops entry is null")
            return nil
        }
        guard (entryRaw & (1 << 63)) != 0 else {
            log("  [-] hook ops entry is not auth-rebase encoded: 0x\(String(format: "%016X", entryRaw))")
            return nil
        }

        let wrapperOff = decodeChainedPtr(entryRaw)
        guard wrapperOff >= 0 else {
            log("  [-] decoded wrapper target invalid: 0x\(String(format: "%X", Int(entryRaw & 0x3FFF_FFFF)))")
            return nil
        }
        let inCode = codeRanges.contains { wrapperOff >= $0.start && wrapperOff < $0.end }
        guard inCode else {
            log("  [-] wrapper target not in code range: 0x\(String(format: "%X", wrapperOff))")
            return nil
        }

        log("  [+] hook cred-label wrapper ops[\(Self.hookCredLabelIndex)] entry 0x\(String(format: "%X", entryOff)) -> 0x\(String(format: "%X", wrapperOff))")
        return (opsTable, entryOff, entryRaw, wrapperOff)
    }

    // MARK: - vfs_context_current Finder

    /// Locate vfs_context_current by its unique 5-word prologue pattern.
    private func findVfsContextCurrentByShape() -> Int {
        let cacheKey = "c23_vfs_context_current"
        if let cached = jbScanCache[cacheKey] { return cached }

        guard let (ks, ke) = kernTextRange else {
            jbScanCache[cacheKey] = -1
            return -1
        }
        let pat = Self.vfsContextCurrentShape
        var hits: [Int] = []

        var off = ks
        while off + pat.count * 4 <= ke {
            var match = true
            for i in 0 ..< pat.count {
                if buffer.readU32(at: off + i * 4) != pat[i] {
                    match = false
                    break
                }
            }
            if match { hits.append(off) }
            off += 4
        }

        let result = hits.count == 1 ? hits[0] : -1
        if result >= 0 {
            log("  [+] vfs_context_current body at 0x\(String(format: "%X", result)) (shape match)")
        } else {
            log("  [-] vfs_context_current shape scan ambiguous (\(hits.count) hits)")
        }
        jbScanCache[cacheKey] = result
        return result
    }

    // MARK: - vnode_getattr Finder

    /// Resolve vnode_getattr from a BL near its log string "vnode_getattr".
    private func findVnodeGetattrViaString() -> Int {
        guard let strOff = buffer.findString("vnode_getattr") else { return -1 }

        // Scan for references to this and nearby instances
        var searchStart = strOff
        for _ in 0 ..< 6 {
            let refs = findStringRefs(searchStart)
            if let ref = refs.first {
                let refOff = ref.adrpOff
                // Python scans backward from the string ref so we prefer the
                // nearest call site rather than the first BL in the window.
                var scanOff = refOff - 4
                let scanLimit = max(0, refOff - 80)
                while scanOff >= scanLimit {
                    let insn = buffer.readU32(at: scanOff)
                    if insn >> 26 == 0b100101 { // BL
                        let imm26 = insn & 0x03FF_FFFF
                        let signedImm = Int32(bitPattern: imm26 << 6) >> 6
                        let target = scanOff + Int(signedImm) * 4
                        let inCode = codeRanges.contains { target >= $0.start && target < $0.end }
                        if inCode {
                            log("  [+] vnode_getattr at 0x\(String(format: "%X", target)) (via BL at 0x\(String(format: "%X", scanOff)))")
                            return target
                        }
                    }
                    scanOff -= 4
                }
            }
            // Try next occurrence
            guard let nextOff = buffer.findString("vnode_getattr", from: searchStart + 1) else { break }
            searchStart = nextOff
        }
        return -1
    }

    // MARK: - C23 Cave Builder

    /// Build the faithful upstream C23 shellcode (exactly 46 instructions = 184 bytes).
    ///
    /// Layout (instruction offsets from caveOff):
    ///   0: nop
    ///   1: cbz x3, #0xa8       (skip if arg3=vp is null)
    ///   2: sub sp, sp, #0x400
    ///   3: stp x29, x30, [sp]
    ///   4: stp x0, x1, [sp, #0x10]
    ///   5: stp x2, x3, [sp, #0x20]
    ///   6: stp x4, x5, [sp, #0x30]
    ///   7: stp x6, x7, [sp, #0x40]
    ///   8: nop
    ///   9: bl vfs_context_current   (position-dependent)
    ///  10: mov x2, x0
    ///  11: ldr x0, [sp, #0x28]     (vp = saved x3)
    ///  12: add x1, sp, #0x80       (vap = &stack_vap)
    ///  13: mov w8, #0x380          (va_supported bitmask)
    ///  14: stp xzr, x8, [x1]      (vap->va_active = 0, vap->va_supported = 0x380)
    ///  15: stp xzr, xzr, [x1, #0x10]
    ///  16: nop
    ///  17: bl vnode_getattr         (position-dependent)
    ///  18: cbnz x0, #0x4c          (skip fixup on error)
    ///  19: mov w2, #0
    ///  20: ldr w8, [sp, #0xcc]     (vap + offset for va_mode bits)
    ///  21: tbz w8, #0xb, #0x14
    ///  22: ldr w8, [sp, #0xc4]
    ///  23: ldr x0, [sp, #0x18]     (new ucred*)
    ///  24: str w8, [x0, #0x18]     (ucred->cr_uid)
    ///  25: mov w2, #1
    ///  26: ldr w8, [sp, #0xcc]
    ///  27: tbz w8, #0xa, #0x14
    ///  28: mov w2, #1
    ///  29: ldr w8, [sp, #0xc8]
    ///  30: ldr x0, [sp, #0x18]
    ///  31: str w8, [x0, #0x28]     (ucred->cr_gid)
    ///  32: cbz w2, #0x14           (if nothing changed, skip P_SUGID)
    ///  33: ldr x0, [sp, #0x20]     (proc*)
    ///  34: ldr w8, [x0, #0x454]
    ///  35: orr w8, w8, #0x100      (set P_SUGID)
    ///  36: str w8, [x0, #0x454]
    ///  37: ldp x0, x1, [sp, #0x10]
    ///  38: ldp x2, x3, [sp, #0x20]
    ///  39: ldp x4, x5, [sp, #0x30]
    ///  40: ldp x6, x7, [sp, #0x40]
    ///  41: ldp x29, x30, [sp]
    ///  42: add sp, sp, #0x400
    ///  43: nop
    ///  44: b wrapperOff             (position-dependent)
    ///  45: nop
    private func buildC23Cave(
        caveOff: Int,
        vfsContextCurrentOff: Int,
        vnodeGetattrOff: Int,
        wrapperOff: Int
    ) -> Data? {
        var code: [Data] = []

        // 0: nop
        code.append(ARM64.nop)
        // 1: cbz x3, #0xa8  (skip entire body = 42 instructions forward)
        code.append(encodeU32(ARM64.c23_cbzX3_0xA8))
        // 2: sub sp, sp, #0x400
        code.append(encodeU32(ARM64.c23_subSP_0x400))
        // 3: stp x29, x30, [sp]
        code.append(encodeU32(ARM64.c23_stpFP_LR))
        // 4: stp x0, x1, [sp, #0x10]
        code.append(encodeU32(ARM64.c23_stpX0X1_0x10))
        // 5: stp x2, x3, [sp, #0x20]
        code.append(encodeU32(ARM64.c23_stpX2X3_0x20))
        // 6: stp x4, x5, [sp, #0x30]
        code.append(encodeU32(ARM64.c23_stpX4X5_0x30))
        // 7: stp x6, x7, [sp, #0x40]
        code.append(encodeU32(ARM64.c23_stpX6X7_0x40))
        // 8: nop
        code.append(ARM64.nop)

        // 9: bl vfs_context_current
        let blVfsOff = caveOff + code.count * 4
        guard let blVfs = encodeBL(from: blVfsOff, to: vfsContextCurrentOff) else { return nil }
        code.append(blVfs)

        // 10: mov x2, x0
        code.append(encodeU32(ARM64.c23_movX2_X0))
        // 11: ldr x0, [sp, #0x28]   (saved x3 = vp)
        code.append(encodeU32(ARM64.c23_ldrX0_sp_0x28))
        // 12: add x1, sp, #0x80
        code.append(encodeU32(ARM64.c23_addX1_sp_0x80))
        // 13: mov w8, #0x380
        code.append(encodeU32(ARM64.c23_movzW8_0x380))
        // 14: stp xzr, x8, [x1]
        code.append(encodeU32(ARM64.c23_stpXZR_X8))
        // 15: stp xzr, xzr, [x1, #0x10]
        code.append(encodeU32(ARM64.c23_stpXZR_XZR_0x10))
        // 16: nop
        code.append(ARM64.nop)

        // 17: bl vnode_getattr
        let blGetAttrOff = caveOff + code.count * 4
        guard let blGetAttr = encodeBL(from: blGetAttrOff, to: vnodeGetattrOff) else { return nil }
        code.append(blGetAttr)

        // 18: cbnz x0, #0x4c  (skip 19 instructions)
        code.append(encodeU32(ARM64.c23_cbnzX0_0x4c))
        // 19: mov w2, #0
        code.append(encodeU32(ARM64.c23_movW2_0))
        // 20: ldr w8, [sp, #0xcc]
        code.append(encodeU32(ARM64.c23_ldrW8_sp_0xcc))
        // 21: tbz w8, #0xb, #0x14  (skip 5 instrs)
        code.append(encodeU32(ARM64.c23_tbzW8_11_0x14))
        // 22: ldr w8, [sp, #0xc4]
        code.append(encodeU32(ARM64.c23_ldrW8_sp_0xc4))
        // 23: ldr x0, [sp, #0x18]
        code.append(encodeU32(ARM64.c23_ldrX0_sp_0x18))
        // 24: str w8, [x0, #0x18]
        code.append(encodeU32(ARM64.c23_strW8_x0_0x18))
        // 25: mov w2, #1
        code.append(encodeU32(ARM64.c23_movW2_1))
        // 26: ldr w8, [sp, #0xcc]
        code.append(encodeU32(ARM64.c23_ldrW8_sp_0xcc))
        // 27: tbz w8, #0xa, #0x14  (skip 5 instrs)
        code.append(encodeU32(ARM64.c23_tbzW8_10_0x14))
        // 28: mov w2, #1
        code.append(encodeU32(ARM64.c23_movW2_1))
        // 29: ldr w8, [sp, #0xc8]
        code.append(encodeU32(ARM64.c23_ldrW8_sp_0xc8))
        // 30: ldr x0, [sp, #0x18]
        code.append(encodeU32(ARM64.c23_ldrX0_sp_0x18))
        // 31: str w8, [x0, #0x28]
        code.append(encodeU32(ARM64.c23_strW8_x0_0x28))
        // 32: cbz w2, #0x14  (skip 5 instrs)
        code.append(encodeU32(ARM64.c23_cbzW2_0x14))
        // 33: ldr x0, [sp, #0x20]
        code.append(encodeU32(ARM64.c23_ldrX0_sp_0x20))
        // 34: ldr w8, [x0, #0x454]
        code.append(encodeU32(ARM64.c23_ldrW8_x0_0x454))
        // 35: orr w8, w8, #0x100
        code.append(encodeU32(ARM64.c23_orrW8_0x100))
        // 36: str w8, [x0, #0x454]
        code.append(encodeU32(ARM64.c23_strW8_x0_0x454))
        // 37: ldp x0, x1, [sp, #0x10]
        code.append(encodeU32(ARM64.c23_ldpX0X1_0x10))
        // 38: ldp x2, x3, [sp, #0x20]
        code.append(encodeU32(ARM64.c23_ldpX2X3_0x20))
        // 39: ldp x4, x5, [sp, #0x30]
        code.append(encodeU32(ARM64.c23_ldpX4X5_0x30))
        // 40: ldp x6, x7, [sp, #0x40]
        code.append(encodeU32(ARM64.c23_ldpX6X7_0x40))
        // 41: ldp x29, x30, [sp]
        code.append(encodeU32(ARM64.c23_ldpFP_LR))
        // 42: add sp, sp, #0x400
        code.append(encodeU32(ARM64.c23_addSP_0x400))
        // 43: nop
        code.append(ARM64.nop)

        // 44: b wrapperOff
        let branchBackOff = caveOff + code.count * 4
        guard let branchBack = encodeB(from: branchBackOff, to: wrapperOff) else { return nil }
        code.append(branchBack)

        // 45: nop
        code.append(ARM64.nop)

        guard code.count == Self.c23CaveWords else {
            log("  [-] C23 cave length drifted: \(code.count) insns, expected \(Self.c23CaveWords)")
            return nil
        }
        return code.reduce(Data(), +)
    }

    // MARK: - Auth-Rebase Pointer Encoder

    /// Retarget an auth-rebase chained pointer while preserving PAC metadata.
    private func encodeAuthRebaseLike(origVal: UInt64, targetFoff: Int) -> Data? {
        guard (origVal & (1 << 63)) != 0 else { return nil }
        // Preserve all bits above bit 31, replace the low 32 bits with target foff
        let newVal = (origVal & ~UInt64(0xFFFF_FFFF)) | (UInt64(targetFoff) & 0xFFFF_FFFF)
        return withUnsafeBytes(of: newVal.littleEndian) { Data($0) }
    }

    // MARK: - Encoding Helper

    private func encodeU32(_ value: UInt32) -> Data {
        ARM64.encodeU32(value)
    }
}
